caikit/runtime/model_management/model_manager.py

```diff
@@ -65,92 +70,230 @@ class ModelManager:
+        try:
+            disk_models = os.listdir(self._local_models_dir)
+        except FileNotFoundError:
+            if self._lazy_sync_timer is not None:
+                self._lazy_sync_timer.cancel()
+                self._lazy_sync_timer = None
+            return
+
+        # Find all models that aren't currently loaded
+        # NOTE: If one of these models gets loaded after this check, that will
+        #   be handled in load_models
+        new_models = [
+            model_id for model_id in disk_models if model_id not in self.loaded_models
+        ]
+        log.debug("New local models: %s", new_models)
+
+        # Find all models that are currently loaded from the local models dir
+        # that no longer exist
+        unload_models = [
+            model_id
+            for model_id, loaded_model in self.loaded_models.items()
+            if model_id not in disk_models
+            and loaded_model.path().startswith(
+                self._local_models_dir,
             )
-            raise CaikitRuntimeException(
-                StatusCode.INTERNAL, "No standalone models loaded"
+        ]
+        log.debug("Unloaded local models: %s", unload_models)
+
+        # Load new models
+        # NOTE: No need for error handling here since load_model will not enter
+        #   failure conditions when not blocking
+        for model_id in new_models:
+            model_path = os.path.join(self._local_models_dir, model_id)
+            self.load_model(model_id, model_path, self._LOCAL_MODEL_TYPE, wait=False)
+
+        # Unload old models
+        # NOTE: No need for error handling here since unload_model will warn on
+        #   errors and move on
+        for model_id in unload_models:
+            log.debug2("Unloading local model %s", model_id)
+            self.unload_model(model_id)
+
+        # Wait for models to load
+        if wait:
+            for model_id in new_models:
+                loaded_model = self.loaded_models.get(model_id)
+                # If somehow already purged, there's nothing to wait on. This is
+                # extremely unlikely since it would require another thread to
+                # explicitly call unload on the model AND have the model finish
+                # loading between then and now. Better to be safe than sorry,
+                # though!
+                if loaded_model is None:  # pragma: no cover
+                    continue
+                # Wait for it and make sure it didn't fail
+                try:
+                    loaded_model.wait()
+                except CaikitRuntimeException as err:
+                    log.warning(
+                        "<RUN56627485W>",
+                        "Failed to load model %s: %s",
+                        model_id,
+                        repr(err),
+                        exc_info=True,
+                    )
+
+        # If running periodically, kick off the next iteration
+        if self._enable_lazy_load_poll:
```

Comment:
Should we stuff the above implementation details into a private function and wrap this up a bit neater into something like:

```
def sync_local_models():

    try:
        self._sync()
    except Exception as e:
        log.warning("stuff went wrong...")
        
    if self._lazy_sync_timer is None:
        ...
    self._lazy_sync_timer.start()
```

I'd like to both:
- make it a bit more clear that this function is responsible for re-invoking itself via the timer
- make really certain that we're wrapping all of the work in an exception handler so that a failure during one iteration doesn't end up killing the bit that queues up the next call

Reply:
Very good point. Let me see how quickly I can bang that out

Reply:
Done